src/charge_controller/dev/ControllerUI.cpp

```diff
@@ -0,0 +1,106 @@
+            setModelState(ControllerModel::State::SETTING_SELECT);
+        }
+        break;
+    case ControllerModel::State::SETTING_SELECT:
+        switch (encoder.getPosition()) {
+        case ControllerModel::SelectedSetting::VOLTAGE:
+            model.setSelectedSetting(ControllerModel::SelectedSetting::VOLTAGE);
+            break;
+        case ControllerModel::SelectedSetting::CURRENT:
+            model.setSelectedSetting(ControllerModel::SelectedSetting::CURRENT);
+            break;
+        case ControllerModel::SelectedSetting::SAVE:
+            model.setSelectedSetting(ControllerModel::SelectedSetting::SAVE);
+            break;
+        case ControllerModel::SelectedSetting::QUIT:
+            model.setSelectedSetting(ControllerModel::SelectedSetting::QUIT);
+            break;
+        case ControllerModel::SelectedSetting::RESET:
+            model.setSelectedSetting(ControllerModel::SelectedSetting::RESET);
+        default:
+            //THIS SHOULD NEVER HAPPEN!!!!
+            break;
+        }
+        if (encoderButton.debounce(DEBOUNCE_TIME)) {
+            switch (encoder.getPosition()) {
+            case ControllerModel::SelectedSetting::VOLTAGE:
+                setModelState(ControllerModel::State::VOLTAGE_SELECT);
+                break;
+            case ControllerModel::SelectedSetting::CURRENT:
+                setModelState(ControllerModel::State::CURRENT_SELECT);
+                break;
+            case ControllerModel::SelectedSetting::SAVE:
+                model.saveVoltageAndCurrent();
+                break;
+            case ControllerModel::SelectedSetting::QUIT:
+                setModelState(ControllerModel::State::PAGE_SELECT);
+                model.setUnsavedVoltage(model.getSavedVoltage());
+                model.setUnsavedCurrent(model.getSavedCurrent());
+                break;
+            case ControllerModel::SelectedSetting::RESET:
+                model.setUnsavedVoltage(DEFAULT_VOLTAGE);
+                model.setUnsavedCurrent(DEFAULT_CURRENT);
+                break;
+            }
+        }
+        break;
+    case ControllerModel::State::VOLTAGE_SELECT:
+        model.setUnsavedVoltage(encoder.getPosition());
+        if (encoderButton.debounce(DEBOUNCE_TIME)) {
+            setModelState(ControllerModel::State::SETTING_SELECT);
+        }
+        break;
+    case ControllerModel::State::CURRENT_SELECT:
+        model.setUnsavedCurrent(encoder.getPosition());
+        if (encoderButton.debounce(DEBOUNCE_TIME)) {
+            setModelState(ControllerModel::State::SETTING_SELECT);
+        }
+        break;
+    }
+    display.display();
+}
+
+void ControllerUI::setModelState(ControllerModel::State newState) {
+    if (model.getState() != newState) {
+        switch (newState) {
+        case ControllerModel::State::SETTING_SELECT:
+            if (model.getState() != ControllerModel::State::PAGE_SELECT) {
```

Comment:
Nitpicky, but you could run `model.getState()` once instead of twice if you store it in a variable at the beginning of the function

Reply:
Fixed!